ts/test/session/sending/MessageQueue_test.ts

```diff
@@ -0,0 +1,202 @@
+    hasSessionStub = sandbox.stub(SessionProtocol, 'hasSession').resolves(true);
+    sandbox.stub(SessionProtocol, 'sendSessionRequest').resolves();
+    sandbox
+      .stub(SessionProtocol, 'sendSessionRequestIfNeeded')
+      .callsFake(async (pubkey: PubKey) => {
+        pubkey;
+        sessionRequestSent = true;
+      });
+
+    // Pending Mesage Cache Stubs
+    const chatMessages = Array.from({ length: 10 }, generateChatMessage);
+    const rawMessage = toRawMessage(
+      generateFakePubkey(),
+      generateChatMessage()
+    );
+
+    sandbox.stub(PendingMessageCache.prototype, 'add').resolves(rawMessage);
+    sandbox.stub(PendingMessageCache.prototype, 'remove').resolves();
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getDevices')
+      .returns(generateMemberList(10));
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getForDevice')
+      .returns(chatMessages.map(m => toRawMessage(generateFakePubkey(), m)));
+
+    messageQueueStub = new MessageQueue();
+  });
+
+  afterEach(() => {
+    TestUtils.restoreStubs();
+    sandbox.restore();
+  });
+
+  it('can send to a single device', async () => {
+    const device = generateFakePubkey();
+    const message = generateChatMessage();
+
+    const promise = messageQueueStub.send(device, message);
+    await expect(promise).to.be.fulfilled;
+  });
+
+  it('can send to many devices', async () => {
+    const devices = generateMemberList(10);
+    const message = generateChatMessage();
+
+    const promise = messageQueueStub.sendMessageToDevices(devices, message);
+    await expect(promise).to.be.fulfilled;
+  });
+
+  it('can send using multidevice', async () => {
+    const device = generateFakePubkey();
+    const message = generateChatMessage();
+
+    const promise = messageQueueStub.sendUsingMultiDevice(device, message);
+    await expect(promise).to.be.fulfilled;
+  });
+
+  it('can send to open group', async () => {
+    const message = generateOpenGroupMessage();
+    const success = await messageQueueStub.sendToGroup(message);
+
+    expect(success).to.equal(true, 'sending to group failed');
+  });
+
+  it('can send to closed group', async () => {
+    const message = generateClosedGroupMessage();
+    const success = await messageQueueStub.sendToGroup(message);
+
+    expect(success).to.equal(true, 'sending to group failed');
+  });
+
+  it('wont send message to empty group', async () => {
+    groupMembersStub.callsFake(async () => generateMemberList(0));
+
+    const message = generateClosedGroupMessage();
+    const response = await messageQueueStub.sendToGroup(message);
+
+    expect(response).to.equal(
+      false,
+      'sendToGroup send a message to an empty group'
+    );
+  });
+
+  it('wont send invalid message type to group', async () => {
+    // Regular chat message should return false
+    const message = generateChatMessage();
+    const response = await messageQueueStub.sendToGroup(message);
+
+    expect(response).to.equal(
```

Comment:
Maybe also check if `sendMessagesToDevices` and `MessageSender.sendToOpenGroup` are not called?